R/utilities-sensitivity-Calculation.R

```diff
@@ -0,0 +1,317 @@
+  parameters <- getAllParametersMatching(
+    parameterPaths,
+    purrr::pluck(simulationResultsBatch, 1L, 1L, "simulation")
+  )
+
+  purrr::map2_dfr(
+    .x = simulationResultsBatch,
+    .y = parameters,
+    .f = ~ .simulationResultsToPKDataFrame(.x, .y)
+  )
+}
+
+#' Extract time-series dataframe from `SimulationResults` object
+#'
+#' @param simResults A **single** instance of `SimulationResults` R6 object.
+#' @param parameter A **single** instance of `Parameter` R6 object.
+#' @inheritParams sensitivityCalculation
+#'
+#' @keywords internal
+#' @noRd
+.simulationResultsToTimeSeriesDataFrame <- function(simulationResults,
+                                                    parameter,
+                                                    outputPaths) {
+  purrr::map_dfr(
+    .x  = simulationResults,
+    .f  = ~ simulationResultsToDataFrame(.x, quantitiesOrPaths = outputPaths),
+    .id = "ParameterFactor"
+  ) %>%
+    dplyr::rename(
+      Concentration = simulationValues,
+      OutputPath    = paths,
+      Dimension     = dimension,
+      Unit          = unit
+    ) %>%
+    .addParameterColumns(parameter) %>%
+    dplyr::select(
+      dplyr::starts_with("Parameter"),
+      Time, Concentration,
+      dplyr::everything(),
+      -c("IndividualId")
+    ) %>%
+    dplyr::arrange(ParameterPath, ParameterFactor)
+}
+
+#' Extract PK parameters dataframe from `Parameter` object
+#'
+#' @inheritParams .simulationResultsToTimeSeriesDataFrame
+#'
+#' @keywords internal
+#' @noRd
+.simulationResultsToPKDataFrame <- function(simulationResults, parameter) {
+  purrr::map_dfr(
+    .x  = simulationResults,
+    .f  = ~ pkAnalysesToDataFrame(calculatePKAnalyses(.x)),
+    .id = "ParameterFactor"
+  ) %>%
+    dplyr::rename(
+      OutputPath       = QuantityPath,
+      PKParameter      = Parameter,
+      PKParameterValue = Value
+    ) %>%
+    .addParameterColumns(parameter) %>%
+    dplyr::group_by(ParameterPath, PKParameter) %>%
+    dplyr::group_modify(.f = ~ .computePercentChange(.)) %>%
+    dplyr::ungroup() %>%
+    dplyr::select(
+      dplyr::starts_with("Parameter"),
+      dplyr::starts_with("PK"),
+      Unit, PercentChangePK,
+      dplyr::everything(),
+      -c("IndividualId")
+    ) %>%
+    dplyr::arrange(ParameterPath, PKParameter, ParameterFactor)
+}
+
+#' @keywords internal
+#' @noRd
+.convertToWide <- function(data) {
+  data %>%
+    tidyr::pivot_wider(
+      names_from  = PKParameter,
+      values_from = c(PKParameterValue, Unit, PercentChangePK, SensitivityPKParameter),
+      names_glue  = "{PKParameter}_{.value}"
+    ) %>%
+    # columns that should not be included in the excel sheets
+    dplyr::select(-c(".rowid")) %>%
+    dplyr::rename_all(~ stringr::str_remove(.x, "_PKParameterValue")) %>%
+    dplyr::rename_all(~ stringr::str_remove(.x, "PK$|PKParameter$")) %>%
+    # all metrics for each parameter should live together
+    dplyr::select(
```

Comment:
Wohaaa whats this? nonono, we totally cannot hardcode that stuff!

Reply:
This is to make sure two things when PK parameters data is saved to spreadsheet:

- the order in which columns corresponding to each PK parameter appear is the same no matter which PK parameters are asked for
- all columns corresponding to any single PK parameter are clubbed together (e.g., "C_max", "C_max_norm", "C_max_Unit"  "C_max_norm_Unit", "C_max_PercentChange", "C_max_norm_PercentChange"  "C_max_Sensitivity" "C_max_norm_Sensitivity" all appear next to each other rather than being dispersed)

``` r
library(esqlabsR)
library(ospsuite)
library(dplyr)

simPath <- system.file("extdata", "Aciclovir.pkml", package = "ospsuite")
simulation <- loadSimulation(simPath)
outputPaths <- "Organism|PeripheralVenousBlood|Aciclovir|Plasma (Peripheral Venous Blood)"
parameterPaths <- c(
  "Aciclovir|Lipophilicity",
  "Applications|IV 250mg 10min|Application_1|ProtocolSchemaItem|Dose",
  "Neighborhoods|Kidney_pls_Kidney_ur|Aciclovir|Glomerular Filtration-GFR|GFR fraction"
)

set.seed(123)
results <- sensitivityCalculation(
  simulation = simulation,
  outputPaths = outputPaths,
  parameterPaths = parameterPaths,
  variationRange = c(0.1, 2, 20),
  pkParameters = NULL
)
#> Following non-standard PK parameters will not be calculated:
#> 

pkDataWide <- results$pkData %>%
  esqlabsR:::.addRowid() %>%
  esqlabsR:::.convertToWide()

names(pkDataWide)
#>  [1] "OutputPath"                         "ParameterPath"                     
#>  [3] "ParameterFactor"                    "ParameterValue"                    
#>  [5] "C_max"                              "C_max_norm"                        
#>  [7] "C_max_Unit"                         "C_max_norm_Unit"                   
#>  [9] "C_max_PercentChange"                "C_max_norm_PercentChange"          
#> [11] "C_max_Sensitivity"                  "C_max_norm_Sensitivity"            
#> [13] "t_max"                              "t_max_Unit"                        
#> [15] "t_max_PercentChange"                "t_max_Sensitivity"                 
#> [17] "AUC_tEnd"                           "AUC_tEnd_norm"                     
#> [19] "AUC_tEnd_Unit"                      "AUC_tEnd_norm_Unit"                
#> [21] "AUC_tEnd_PercentChange"             "AUC_tEnd_norm_PercentChange"       
#> [23] "AUC_tEnd_Sensitivity"               "AUC_tEnd_norm_Sensitivity"         
#> [25] "AUC_inf"                            "AUC_inf_norm"                      
#> [27] "AUC_inf_Unit"                       "AUC_inf_norm_Unit"                 
#> [29] "AUC_inf_PercentChange"              "AUC_inf_norm_PercentChange"        
#> [31] "AUC_inf_Sensitivity"                "AUC_inf_norm_Sensitivity"          
#> [33] "CL"                                 "FractionAucLastToInf"              
#> [35] "CL_Unit"                            "FractionAucLastToInf_Unit"         
#> [37] "CL_PercentChange"                   "FractionAucLastToInf_PercentChange"
#> [39] "CL_Sensitivity"                     "FractionAucLastToInf_Sensitivity"  
#> [41] "MRT"                                "MRT_Unit"                          
#> [43] "MRT_PercentChange"                  "MRT_Sensitivity"                   
#> [45] "Thalf"                              "Thalf_Unit"                        
#> [47] "Thalf_PercentChange"                "Thalf_Sensitivity"                 
#> [49] "Vss"                                "Vss_Unit"                          
#> [51] "Vss_PercentChange"                  "Vss_Sensitivity"                   
#> [53] "Vd"                                 "Vd_Unit"                           
#> [55] "Vd_PercentChange"                   "Vd_Sensitivity"
```

<sup>Created on 2022-02-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

I will try to think of a way where this doesn't require hardcoding, but this is the best I can do for now.

Reply:
LOL, never mind. Figured out a way to do this without hardcoding. _dplyr_ is awesome. 🔥